tools/cas/cas.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/cachetools"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc/metadata"
+
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
+)
+
+var (
+	// Required flags:
+
+	target     = flag.String("target", "", "Cache grpc target, such as grpcs://remote.buildbuddy.io")
+	blobDigest = flag.String("digest", "", "Digest of the blob to fetch, in HASH/SIZE format.")
+	blobType   = flag.String("type", "", "Type of blob to inspect: Action, ActionResult, Command, file, stdout, stderr")
+
+	// Optional flags:
+
+	instanceName = flag.String("remote_instance_name", "", "Remote instance name")
+	apiKey       = flag.String("api_key", "", "API key to attach to the outgoing context")
+	invocationID = flag.String("invocation_id", "", "Invocation ID. This is required when fetching the result of a failed action. Otherwise, it's optional.")
+)
+
+// Examples:
+//
+// Show an action result proto:
+//     bazel run //tools/cas -- -target=grpcs://remote.buildbuddy.dev -digest=HASH/SIZE -type=ActionResult
+//
+// Show a command proto:
+//     bazel run //tools/cas -- -target=grpcs://remote.buildbuddy.dev -digest=HASH/SIZE -type=Command
+//
+// Show stderr contents:
+//     bazel run //tools/cas -- -target=grpcs://remote.buildbuddy.dev -digest=HASH/SIZE -type=stderr
+//
+// Show a failed action result proto (requires invocation ID):
+//     bazel run //tools/cas -- -target=grpcs://remote.buildbuddy.dev -digest=HASH/SIZE -type=ActionResult -invocation_id=IID
+func main() {
+	flag.Parse()
+	if *target == "" {
+		log.Fatalf("Missing --target")
+	}
+	parts := strings.Split(*blobDigest, "/")
+	if len(parts) != 2 {
+		log.Fatalf("Invalid --digest format: expecting {hash}/{size}, got %q", *blobDigest)
+	}
+	sizeBytes, err := strconv.Atoi(parts[1])
+	if err != nil {
+		log.Fatalf("Error parsing digest: %s", err)
+	}
+	ind := digest.NewResourceName(&repb.Digest{
+		Hash:      parts[0],
+		SizeBytes: int64(sizeBytes),
+	}, *instanceName)
+	conn, err := grpc_client.DialTarget(*target)
+	if err != nil {
+		log.Fatalf("Error dialing CAS target: %s", err.Error())
+	}
+	bsClient := bspb.NewByteStreamClient(conn)
+	acClient := repb.NewActionCacheClient(conn)
+
+	ctx := context.Background()
+	if *apiKey != "" {
+		ctx = metadata.AppendToOutgoingContext(ctx, "x-buildbuddy-api-key", *apiKey)
```

Comment:
Would be nice to add a utility function for this. We've been doing this in a bunch of different tools.

Reply:
I agree -- would prefer do this as a separate PR though, probably at a later date when I have more time to work on CLI tools :)

Reply:
:+1: